### ISSUE TITLE
About interruption of generic checkers

### ISSUE THREAD

**Issue by Sarcasm** [flycheck/flycheck#585]
Hi,

I'm debugging flycheck-irony which was buggy from the beginning...

I now have more information at my disposal to fulfill the [Status callback protocol](http://flycheck.readthedocs.org/en/latest/dev/api.html#status-callback-protocol), before that I wasn't respecting the protocol and was falling into what the warning that says "Failure to call the callback will cause Flycheck to get stuck at the current syntax check.".

I have an issue with the way the `interrupt` status is handled, maybe that should be the same for `errored`.

For the purpose of this example, let's assume syntax checking is taking a constant time to be performed of 3 seconds.

My initial C++ code is like this, the pipe character `|` representing the position of the cursor:

```cpp
struct foo {}|
```

The code is faulty, missing a semicolon `;` after the closing brace `}`.

So the first syntax checking happens after 3 seconds and the error is reported.

        error: expected ';' after struct... (irony)

Now if I do the following steps:

1. change the buffer, i.e: adding `;`
2. flycheck start the syntax checking, it will take 3 seconds to complete
3. I revert my change 1 second after syntax checking has started
4. The syntax checking that was running finally completes but since `(buffer-chars-modified-tick)` has changed compared to the initial request, I consider the change as "obselete", in this case I give the `interrupted` status to the callback
5. Now the buffer is still incomplete `struct foo {}` but flycheck doesn't say so.

I would expect an interrupted syntax checking to not stop further checking if the buffer has changed, do you agree?

Should my checker re-trigger a syntax checking request internally when I detect such buffer change?

**Comment by lunaryorn** [comment]
@Sarcasm I'm not sure whether I understand you correctly:  Does your syntax checker really interrupt itself if it has detected a change to the buffer?  If yes, just don't do that.

**Comment by Sarcasm** [comment]
I'm not sure it interrupts itself in the sense you are thinking about.

When I get a syntax check request it is actually decomposed in 2 steps:
1. parse the buffer context (this is the one that takes 3 seconds)
2. if the parsing is successful and still in sync with the buffer (that's where `(buffer-chars-modified-tick)` is checked) then I send the second requests, that returns the diagnostics. If that is not in sync I avoid this second request and report that I was "interrupted".

Is any of that wrong?

**Comment by Sarcasm** [comment]
Nevermind, if I return a `finished` status with 0 errors flycheck takes into account the buffer change and make another request immediately, that's all I need.

I guess I don't understand exactly what `interrupt` is for but I don't seem to need it.

**Comment by lunaryorn** [comment]
@Sarcasm `interrupt` is for forced interruptions from the outside, i.e. if a buffer is killed or reverted, or if the user forcibly interrupts a syntax check.  It's a status you'll only return if you have an `:interrupt-fn` in your definition.

**Comment by Sarcasm** [comment]
Thanks for the explanation on `interrupted`, I don't think I need it in my case.

Regarding the early exit, I think you are worrying too much here, I'm pretty sure I'm not abusing flycheck in any way. This is not really done for the purpose of optimization but also correctness, I just can't let irony give information about an old parsing because the information may no longer be available already.

I'm actually trying to live by the rules that flycheck dictates, I really want to avoid hacks and I think it's the same for you.
